src/machine/state/mem.rs

```diff
@@ -1,43 +1,104 @@
 use machine::state::types::*;
 
+use std::mem::transmute;
 use std::ops::{Index, IndexMut};
 
 pub struct ByteAt(pub u64);
 pub struct WydeAt(pub u64);
 pub struct TetraAt(pub u64);
 pub struct OctaAt(pub u64);
 
-pub struct Memory(());
+pub struct Memory(Vec<u8>);
+
+// 2^30
+// TODO let the user choose the size
+const MAX_ADDRESS : usize = 0x40000000;
+
+static ZERO_BYTE : Byte = Byte(0);
+static ZERO_WYDE: Wyde = Wyde(0);
 
 impl Memory {
     pub fn new() -> Self {
-        unimplemented!();
+        Memory(Vec::new())
     }
 }
 
 impl Index<ByteAt> for Memory {
     type Output = Byte;
-    fn index(&self, _: ByteAt) -> &Self::Output {
-        unimplemented!();
+    fn index(&self, index: ByteAt) -> &Self::Output {
+        let i = index.0 as usize;
+        assert!(i < MAX_ADDRESS);
+        println!("{} >= {}?", i, self.0.len());
+        if i >= self.0.len() {
+            &ZERO_BYTE
+        } else {
+            let raw_ref = self.0.get(i).unwrap();
+            unsafe {
+                transmute::<&u8, &Byte>(raw_ref)
+            }
+        }
     }
 }
 
 impl IndexMut<ByteAt> for Memory {
-    fn index_mut(&mut self, _: ByteAt) -> &mut Self::Output {
-        unimplemented!();
+    fn index_mut(&mut self, index: ByteAt) -> &mut Self::Output {
+        let i = index.0 as usize;
+        assert!(i < MAX_ADDRESS);
+        if i >= self.0.len() {
+            // make the new size a multiple of 8, so that no problems arise for
+            // Wydes, Tetras and Octas
+            let new_size = (i | 111) + 1;
+            self.0.resize(new_size, 0);
+        }
+        let raw_ref = self.0.get_mut(i).unwrap();
+        unsafe {
+            transmute::<&mut u8, &mut Byte>(raw_ref)
+        }
     }
 }
 
 impl Index<WydeAt> for Memory {
     type Output = Wyde;
-    fn index(&self, _: WydeAt) -> &Self::Output {
-        unimplemented!();
+    fn index(&self, index: WydeAt) -> &Self::Output {
+
+        let lo = (index.0 | 1) as usize;
+        let hi = lo - 1;
+
+        assert!(lo < MAX_ADDRESS);
+
+        if lo >= self.0.len() {
+            &ZERO_WYDE
+        } else {
+            println!("vector length: {}", self.0.len());
+            println!("requested address: {}", index.0);
+            println!("high byte at index {}: {}", hi, self.0[hi]);
+            println!("low byte at index {}: {}", lo, self.0[lo]);
+            unsafe {
+                transmute::<&(u8, u8), &Wyde>(&(self.0[hi], self.0[lo]))
```

Comment:
This is an awesome example of why `transmute` is dangerous. Not only did it transmute the pointee-type of the reference, but also the lifetime. You pass in a reference of a short lifetime but (through type inference) get out a reference of long lifetime. Dangerous³.